### ISSUE TITLE
zpool labelclear on full device does not clear label on part1

### ISSUE THREAD

**Issue by jimsalterjrs** [zfsonlinux/zfs#3156]
Very confusing behavior when feeding full disks to a zpool.  

TL;DR you aren't really using a full disk, you're letting ZFS partition the full disk and use partition 1 - but it doesn't really handle all operations properly when referring to the full disk afterward.

Expected behavior: ZFS should handle references to full disks consistently.
Current behavior: references to full disks sometimes work, and sometimes don't.

Example:

````
root@thumb:~# zpool status
  pool: test
 state: ONLINE
  scan: resilvered 10.0G in 0h10m with 0 errors on Thu Mar  5 14:04:14 2015
config:

	NAME                        STATE     READ WRITE CKSUM
	test                        ONLINE       0     0     0
	  raidz2-0                  ONLINE       0     0     0
	    wwn-0x50014ee20589576f  ONLINE       0     0     0
	    wwn-0x50014ee259418099  ONLINE       0     0     0
	    wwn-0x50014ee259481bfe  ONLINE       0     0     0
	    wwn-0x50014ee2594942f9  ONLINE       0     0     0
	    wwn-0x50014ee2ae96ed1a  ONLINE       0     0     0
	    wwn-0x50014ee2b6275ac1  ONLINE       0     0     0

errors: No known data errors
root@thumb:~# zpool offline test wwn-0x50014ee2b6275ac1
root@thumb:~# zpool export test
root@thumb:~# zpool labelclear -f /dev/disk/by-id/wwn-0x50014ee2b6275ac1
root@thumb:~# zpool import test
root@thumb:~# zpool status
  pool: test
 state: DEGRADED
status: One or more devices has been taken offline by the administrator.
	Sufficient replicas exist for the pool to continue functioning in a
	degraded state.
action: Online the device using 'zpool online' or replace the device with
	'zpool replace'.
  scan: resilvered 10.0G in 0h10m with 0 errors on Thu Mar  5 14:04:14 2015
config:

	NAME                        STATE     READ WRITE CKSUM
	test                        DEGRADED     0     0     0
	  raidz2-0                  DEGRADED     0     0     0
	    wwn-0x50014ee20589576f  ONLINE       0     0     0
	    wwn-0x50014ee259418099  ONLINE       0     0     0
	    wwn-0x50014ee259481bfe  ONLINE       0     0     0
	    wwn-0x50014ee2594942f9  ONLINE       0     0     0
	    wwn-0x50014ee2ae96ed1a  ONLINE       0     0     0
	    wwn-0x50014ee2b6275ac1  OFFLINE      0     0     0

errors: No known data errors
````

So far, so good... apparently.  But now it goes haywire.

````
root@thumb:~# zpool replace test wwn-0x50014ee2b6275ac1 wwn-0x50014ee2b6275ac1
invalid vdev specification
use '-f' to override the following errors:
/dev/disk/by-id/wwn-0x50014ee2b6275ac1 contains a corrupt primary EFI label.
root@thumb:~# zpool replace -f test wwn-0x50014ee2b6275ac1 wwn-0x50014ee2b6275ac1
cannot replace wwn-0x50014ee2b6275ac1 with wwn-0x50014ee2b6275ac1: wwn-0x50014ee2b6275ac1 is busy
````

Lol wut?  Quite a lot of experimentation later, I realized that while ZFS doesn't ever *tell* you so, it's *really* using partition 1 - so that label needs to be cleared.

````
root@thumb:~# zpool labelclear -f /dev/disk/by-id/wwn-0x50014ee2b6275ac1-part1
labelclear operation failed.
	Vdev /dev/disk/by-id/wwn-0x50014ee2b6275ac1-part1 is a member (ACTIVE), of pool "test".
	To remove label information from this device, export or destroy
	the pool, or remove /dev/disk/by-id/wwn-0x50014ee2b6275ac1-part1 from the configuration of this pool
	and retry the labelclear operation.
root@thumb:~# zpool export test
root@thumb:~# zpool labelclear -f /dev/disk/by-id/wwn-0x50014ee2b6275ac1-part1
root@thumb:~# zpool import test
root@thumb:~# zpool replace -f test wwn-0x50014ee2b6275ac1 wwn-0x50014ee2b6275ac1
````

And now things are working.

**Comment by kpande** [comment]
this could be somewhat related to the issues in #2844 #2876 #3145

**Comment by joshenders** [comment]
Noticing similar odd `zpool labelclear` behavior. May be related, if not I'll create a new issue. It appears that a `zpool replace` is recreating zfs partitions from metadata that `zpool labelclear` failed to remove? maybe I need to `zpool labelclear` the `-part1` id and not the full device?

```
root@collaris: ~ # ls -lash /dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F3WXSM*
0 lrwxrwxrwx 1 root root  9 Apr 24 16:49 /dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F3WXSM -> ../../sdj
0 lrwxrwxrwx 1 root root 10 Apr 24 16:49 /dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F3WXSM-part1 -> ../../sdj1
0 lrwxrwxrwx 1 root root 10 Apr 24 16:49 /dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F3WXSM-part9 -> ../../sdj9
root@collaris: ~ # zpool labelclear -f /dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F3WXSM
root@collaris: ~ # ls -lash /dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F3WXSM*
0 lrwxrwxrwx 1 root root 9 Apr 24 16:49 /dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F3WXSM -> ../../sdj
root@collaris: ~ # zpool replace -f data ata-ST3000DM001-9YN166_Z1F11382 /dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F3WXSM
cannot replace ata-ST3000DM001-9YN166_Z1F11382 with /dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F3WXSM: /dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F3WXSM is busy
root@collaris: ~ # ls -lash /dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F3WXSM*
0 lrwxrwxrwx 1 root root  9 Apr 24 16:49 /dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F3WXSM -> ../../sdj
0 lrwxrwxrwx 1 root root 10 Apr 24 16:49 /dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F3WXSM-part1 -> ../../sdj1
0 lrwxrwxrwx 1 root root 10 Apr 24 16:49 /dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F3WXSM-part9 -> ../../sdj9
```

**Comment by aj-git** [comment]
Just testet zpool labelclear decommissioning a mirror-pool @ Ubuntu 14.04 using
```
# dmesg | grep -E 'SPL:|ZFS:'
	SPL: Loaded module v0.6.4.2-1~trusty
	ZFS: Loaded module v0.6.4.2-1~trusty, ZFS pool version 5000, ZFS filesystem version 5
```
To sum it up: after zpool labelclear the kernel does not automatically refresh the partition table of the disks. Manually forcing a refresh of the partition table for each label-cleared disk was necessary:
```
blockdev --rereadpt /dev/sdN
or
blockdev --rereadpt /dev/disk/by-id/ata-...
```
(for me /dev/disk/by-id/ata-...-part1 & /dev/disk/by-id/ata-...-part9 disappeared)

Afterwards a "zpool create -f ..." allowed me to create a new pool without any problems.

**Comment by ilovezfs** [comment]
Are we sure this should actually be considered a bug as opposed to a documentation issue? It seems to me that there is a loss of flexibility if labelclear /dev/sda goes with /dev/sda1 instead since both are possible locations for a pool, and reading whole_disk=0 (should clear /dev/sda) vs. whole_disk=1 (should clear /dev/sda1) may not always be possible. Or to put it another way, if labelclear uses sda1 automatically and not the exact block device specified, how would you actually perform a labelclear on sda itself?

**Comment by CMCDragonkai** [comment]
I may have a similar situation, I needed to change the partitions of the disk and basically wanted a fresh start on my disk. Using `zpool labelclear` and `sgdisk --zap-all` was not sufficient to clear the disk entirely of ZFS metadata. After setting up the zpool again, the system failed to boot complaining that there was 2 zfs things (forgot what the word was).

Instead I have to zero out the disk entirely for it to work using `dd``.

**Comment by devZer0** [comment]
i wanted to convert a virtual scsi disk to kvm virtio, so i detached scsi6 from the kvm in proxmox and hot re-added the same qcow2 file as virtio6.   

it re-appeared inside vm as "/dev/vda"

i'm not able to re-add that to the existing pool as labelclear won't let me remove the pool information - not even when using -force

# zpool status
  pool: zfspool
 state: DEGRADED
status: One or more devices has experienced an unrecoverable error.  An
	attempt was made to correct the error.  Applications are unaffected.
action: Determine if the device needs to be replaced, and clear the errors
	using 'zpool clear' or replace the device with 'zpool replace'.
   see: http://zfsonlinux.org/msg/ZFS-8000-9P
  scan: scrub repaired 0B in 0 days 02:31:34 with 0 errors on Wed Jul 17 12:39:32 2019
config:

	NAME                                      STATE     READ WRITE CKSUM
	zfspool                                   DEGRADED     0     0     0
	  raidz2-0                                DEGRADED     0     0     0
	    scsi-0QEMU_QEMU_HARDDISK_drive-scsi1  ONLINE       0     0     0
	    scsi-0QEMU_QEMU_HARDDISK_drive-scsi2  ONLINE       0     0     0
	    scsi-0QEMU_QEMU_HARDDISK_drive-scsi3  ONLINE       0     0     0
	    scsi-0QEMU_QEMU_HARDDISK_drive-scsi4  ONLINE       0     0     0
	    scsi-0QEMU_QEMU_HARDDISK_drive-scsi5  ONLINE       0     0     0
	    scsi-0QEMU_QEMU_HARDDISK_drive-scsi6  OFFLINE      3     4     0

errors: No known data errors

# zpool replace zfspool scsi-0QEMU_QEMU_HARDDISK_drive-scsi6 /dev/vda
invalid vdev specification
use '-f' to override the following errors:
/dev/vda1 is part of active pool 'zfspool'

# zpool replace -f zfspool scsi-0QEMU_QEMU_HARDDISK_drive-scsi6 /dev/vda
invalid vdev specification
the following errors must be manually repaired:
/dev/vda1 is part of active pool 'zfspool'

# zpool labelclear /dev/vda
failed to clear label for /dev/vda

# zpool labelclear /dev/vda1
/dev/vda1 is a member (ACTIVE) of pool "zfspool"

# zpool labelclear -f /dev/vda1
/dev/vda1 is a member (ACTIVE) of pool "zfspool"

**Comment by devZer0** [comment]
related:
https://github.com/zfsonlinux/zfs/issues/2076

**Comment by devZer0** [comment]
i really really hate automated ticket closing bots. 

in 20 year's of IT i have stumbled upon dozens of closed but unresolved issues and that really sucks. 

it's like "oh, let's brush it under the carpet if nobody cares or nobody complains....there are so many other things to do..."

See https://github.com/OWASP/glue/issues/111 for discussion on stale-bot usage

**Comment by devZer0** [comment]
this is a "please leave this open until resolved post"

**Comment by racerxdl** [comment]
Bump, same problem here. Need to dd the entire disk with zero so ZFS can replace the disk.

**Comment by jimsalterjrs** [comment]
> Bump, same problem here. Need to dd the entire disk with zero so ZFS can replace the disk.
> 
> +1 to devZer0 comment

You don't need to dd the entire disk unless you've destroyed the partition table already. ASSUMING THE PARTITION TABLE IS STILL INTACT, you can use the `wipefs` command on each partition, then use it on the entire disk.

If you've already destroyed the partition table and aren't able to recreate it exactly as it was, then yeah, unfortunately you'll probably have to wipe the disk; I don't really know HOW it works but I've seen ZFS find a label on a deleted partition and refuse to use the disk before.

**Comment by racerxdl** [comment]
> > Bump, same problem here. Need to dd the entire disk with zero so ZFS can replace the disk.
> > 
> > +1 to devZer0 comment
> 
> You don't need to dd the entire disk unless you've destroyed the partition table already. ASSUMING THE PARTITION TABLE IS STILL INTACT, you can use the `wipefs` command on each partition, then use it on the entire disk.
> 
> If you've already destroyed the partition table and aren't able to recreate it exactly as it was, then yeah, unfortunately you'll probably have to wipe the disk; I don't really know HOW it works but I've seen ZFS find a label on a deleted partition and refuse to use the disk before.

I tried the wipefs -a in the device, it confirmed wiped everythin zfs related. It didnt worked. It said the disk was in use by the pool. The I tried nuke the gpt partiton table, it didnt worked as well.

I'm now dding the whole disk with zero.

**Comment by racerxdl** [comment]
Yeah, only dd'ing zeros to the disk actually fixed it. I'm not sure why.
One thing I noticed it: After wipefs and partition table wipe if I create a single partition on the disk, save, and reopen gparted, the partition type is changed to zfs, which probably means the zfs kernel module is still book keeping the disk even after all clear commands ran. 

For some reason dd'ing the entire disk solve it, but its a real pain since for big disks it takes a VERY LONG time.

**Comment by sbsaylors** [comment]
I couldn't do  the /dev/sdc, but I could do it like this. I was able to them add the drive back and it started to resilver with out any chksum errors.

zpool offline  ZFSonProx /dev/disk/by-id/ata-OOS12000G_0005BAD <---this is /dev/sdc

zpool remove ZFSonProx /dev/disk/by-id/ata-OOS12000G_0005BAD0

zpool labelclear  -f /dev/sdc9
zpool labelclear  -f /dev/sdc1

zpool replace ZFSonProx /dev/disk/by-id/ata-OOS12000G_0005BAD0 /dev/disk/by-id/ata-OOS12000G_0005BAD0

**Comment by bmbeverst** [comment]
Also had this same issue, @joshenders your original comment worked. Even using a second computer with `sgdisk --clear` didn't work, `labelclear` failed as well.
